File: ios/source/store/AppState.swift
import Foundation

struct AppState {
    var connected: ConnectionType? = .none
    var scanning = false
    var servers: [Server] = []
    var projects: [Bloop_ProjectInfo] = []
    var cloudProjects: [Bloop_ProjectInfo] = []
    var project = emptyProject()
    var projectInfo: Bloop_ProjectInfo? = .none
    var projectSyncStatuses: [String: Bloop_SyncStatus] = [:]
    var playbackState = Bloop_PlaybackState()
    var progress = Bloop_Progress()
    var waveforms: [Id: Bloop_WaveformData] = [:]
    var user: Bloop_User? = .none
    var errors: [String] = []
    var preferences: Bloop_Preferences? = .none
}

func emptyProject() -> Bloop_Project {
    Bloop_Project.with {
        $0.songs = []
        $0.selections = Bloop_Selections()
    }
}

File: android/app/src/main/java/com/joenoel/bloop/state/AppState.kt
package com.joenoel.bloop.state

import bloop.Bloop

enum class ConnectionType {
    LOCAL,
    REMOTE,
}

sealed interface ServerEndpoint {
    data class HostPort(val host: String, val port: Int) : ServerEndpoint
    data class Service(
        val name: String,
        val type: String,
        val domain: String,
        val interfaceName: String? = null,
    ) : ServerEndpoint
    data class Url(val value: String) : ServerEndpoint
    data class Opaque(val value: String) : ServerEndpoint
}

data class AppState(
    val connected: ConnectionType? = null,
    val scanning: Boolean = false,
    val servers: List<ServerEndpoint> = emptyList(),
    val projects: List<Bloop.ProjectInfo> = emptyList(),
    val cloudProjects: List<Bloop.ProjectInfo> = emptyList(),
    val project: Bloop.Project = emptyProject(),
    val projectInfo: Bloop.ProjectInfo? = null,
    val projectSyncStatuses: Map<String, Bloop.SyncStatus> = emptyMap(),
    val playbackState: Bloop.PlaybackState = Bloop.PlaybackState.getDefaultInstance(),
    val progress: Bloop.Progress = Bloop.Progress.getDefaultInstance(),
    val waveforms: Map<Long, Bloop.WaveformData> = emptyMap(),
    val user: Bloop.User? = null,
    val errors: List<String> = emptyList(),
    val preferences: Bloop.Preferences? = null,
    val lastResponseText: String? = null,
)

fun emptyProject(): Bloop.Project {
    return Bloop.Project
        .newBuilder()
        .setSelections(Bloop.Selections.getDefaultInstance())
        .build()
}
